googleapis/google-cloud-php
# feat(PubSub): Enable publisher compression

Adding gRPC compression support in publisher.
> b/277153658

#### Implementation Summary
* Main logic resides in `Grpc.php > publishMessage()` method because we need to calculate the total size of the messages to be send after they have been converted into Protobuf message class equivalent.
* Topic.php hosts the logic to set the value of the compression options to be passed down to the Grpc layer.
* BatchPublisher just set's the supplied option in it's variables and then passed them down to the Topic constructor when required.

#### Steps to test this feature in prod
* Clone this branch
* Copy these two scripts in the `PubSub > src` directory ([testRunner.php](https://gist.github.com/yash30201/a06725b0bcaeebe42368eb5bd70f4462) and [prodTest](https://gist.github.com/yash30201/b22438786537e322b96c519a2df6516f)).
* Move into `PubSub/src`.
* Run `php testRunner.php`.

These scripts set the grpc debugging flags, run the required test script(`prodTest.php`), fetch the terminal output due to running of `prodTest.php`, parses the output to check if compression happened or not (_GRPC logs statement like this when it compresses request: `Compressed[gzip] 339 bytes vs. 258 bytes (23.89% savings)`_) and check with the expected output to print the verdict.

More info in design doc's section [PROD TESTING](https://docs.google.com/document/d/1fHiCE9erqD5nM72T6Z_LATqMM6r5NAvnZyQM34BK0Q4/edit#heading=h.91kqo9c082ue)
